### PR TITLE
Fix3.0/issue 320

### PR DESCRIPTION
**Pull request by rgulrajani**
### Description

Implement missing attributes and sub elements for MediaFile

### Issue

https://github.com/dailymotion/vast-client-js/issues/320

### Type
- [ ] Breaking change
- [x] Enhancement
- [x] Fix
- [x] Documentation
- [ ] Tooling

### CODE REVIEW

File: docs/api/class-reference.md [hunk 1]
```diff
@@ -44,6 +44,8 @@ This object represents a generic Creative. It's used as a parent object for more
 - `skipDelay: Number|null`
 - `mediaFiles: Array<Object>` [go to object](#mediafile)
 - `mezzanine: Object<Object>` [go to object](#mezzanine)
+- `interactiveCreativeFile: Object<Object>` [go to object](#interactivecreativefile)
```
Comment (kobawan) [inline docs/api/class-reference.md]: `Object<Object>` should just be `Object`

File: docs/api/class-reference.md [hunk 2]
```diff
@@ -91,6 +95,19 @@ This object represents a generic Creative. It's used as a parent object for more
 - `mediaType: String|null` Type of media file (3D / 360 / etc). Optional. Default value = 2D
 - `fileURL: String|null`
 
+## InteractiveCreativeFile<a name="interactivecreativefile"></a>
+
+- `type: String` MIME type for the file provided
+- `apiFramework: String` Identifies the API needed to execute the resource file, if applicable
+- `variableDuration: Boolean`  Identifies whether the ad always drops when the duration is reached, or if it can potentially extend the duration by pausing the underlying video or delaying the adStopped call after adVideoComplete
+- `fileURL: String|null`
```
Comment (kobawan) [inline docs/api/class-reference.md]: it would be great to specify which values could be optional (`null`)

File: src/interactive_creative_file.js [hunk 3]
```diff
@@ -0,0 +1,8 @@
+export function createInteractiveCreativeFile() {
+  return {
```
Comment (kobawan) [inline src/interactive_creative_file.js]: you could already pass the properties as params in the function

File: src/parser/creative_linear_parser.js [hunk 4]
```diff
@@ -206,69 +192,151 @@ export function parseCreativeLinear(creativeElement, creativeAttributes) {
...
+  );
+  let variableDuration = interactiveCreativeElement.getAttribute(
+    'variableDuration'
+  );
+  if (variableDuration && typeof variableDuration === 'string') {
+    variableDuration = variableDuration.toLowerCase();
+    if (variableDuration === 'true') {
+      interactiveCreativeFile.variableDuration = true;
+    } else if (variableDuration === 'false') {
+      interactiveCreativeFile.variableDuration = false;
```
Comment (kobawan) [inline src/parser/creative_linear_parser.js]: don't we have a util to correctly parse boolean variables? If not you could make one as it is used in multiple places now

File: src/interactive_creative_file.js (hunk 3 above)
Comment (rgulrajani) [reply src/interactive_creative_file.js]: Done.

File: docs/api/class-reference.md (hunk 2 above)
Comment (rgulrajani) [reply docs/api/class-reference.md]: Done

File: src/parser/creative_linear_parser.js (hunk 4 above)
Comment (rgulrajani) [reply src/parser/creative_linear_parser.js]: parseBoolean exists. And used

File: src/interactive_creative_file.js [hunk 5]
```diff
@@ -1,8 +1,15 @@
...
   return {
-    type: null,
-    apiFramework: null,
-    variableDuration: null,
+    type: interactiveCreativeAttributes.type || null,
+    apiFramework: interactiveCreativeAttributes.apiFramework || null,
+    variableDuration:
+      parserUtils.parseBoolean(
+        interactiveCreativeAttributes.variableDuration
+      ) || null,
```
Comment (kobawan) [inline src/interactive_creative_file.js]: here it will result in `null` if value is false

File: src/interactive_creative_file.js [hunk 6]
```diff
@@ -1,8 +1,15 @@
-export function createInteractiveCreativeFile() {
+import { parserUtils } from './parser/parser_utils';
+
+export function createInteractiveCreativeFile(
+  interactiveCreativeAttributes = {}
```
Comment (kobawan) [inline src/interactive_creative_file.js]: just a suggestion, you could destructure the object already here :)

File: src/closed_caption_file.js [hunk 7]
```diff
@@ -0,0 +1,7 @@
+export function createClosedCaptionFile() {
```
Comment (kobawan) [inline src/closed_caption_file.js]: same for the params in this function

File: src/parser/creative_linear_parser.js [hunk 8]
```diff
@@ -206,69 +192,136 @@ export function parseCreativeLinear(creativeElement, creativeAttributes) {
...
+
+  let scalable = mediaFileElement.getAttribute('scalable');
+  if (scalable && typeof scalable === 'string') {
+    scalable = scalable.toLowerCase();
+    if (scalable === 'true') {
+      mediaFile.scalable = true;
+    } else if (scalable === 'false') {
+      mediaFile.scalable = false;
+    }
+  }
```
Comment (kobawan) [inline src/parser/creative_linear_parser.js]: please re-use the boolean function :)

File: src/parser/creative_linear_parser.js [hunk 9]
```diff
@@ -206,69 +192,136 @@ export function parseCreativeLinear(creativeElement, creativeAttributes) {
...
+    'maintainAspectRatio'
+  );
+  if (maintainAspectRatio && typeof maintainAspectRatio === 'string') {
+    maintainAspectRatio = maintainAspectRatio.toLowerCase();
+    if (maintainAspectRatio === 'true') {
+      mediaFile.maintainAspectRatio = true;
+    } else if (maintainAspectRatio === 'false') {
+      mediaFile.maintainAspectRatio = false;
+    }
+  }
```
Comment (kobawan) [inline src/parser/creative_linear_parser.js]: same here with the boolean function :)

File: src/parser/creative_linear_parser.js (hunk 8 above)
Comment (rgulrajani) [reply src/parser/creative_linear_parser.js]: Done

File: src/interactive_creative_file.js (hunk 6 above)
Comment (rgulrajani) [reply src/interactive_creative_file.js]: I think I am going to keep it as an object for now. Also for consistency with the creativeAttributes which are passed in the companionAd

File: src/interactive_creative_file.js (hunk 5 above)
Comment (rgulrajani) [reply src/interactive_creative_file.js]: Fixed

File: src/closed_caption_file.js (hunk 7 above)
Comment (rgulrajani) [reply src/closed_caption_file.js]: Done

**Pull request merged**